File: cotgenerator/src/main/java/com/jon/cotgenerator/GeneratorCotFactory.java
package com.jon.cotgenerator;

import android.content.Context;
import android.content.SharedPreferences;

import com.jon.common.cot.CotRole;
import com.jon.common.cot.CotTeam;
import com.jon.common.cot.CursorOnTarget;
import com.jon.common.cot.UtcTimestamp;
import com.jon.common.service.CotFactory;
import com.jon.common.service.Point;
import com.jon.common.utils.Constants;
import com.jon.common.utils.DeviceUid;
import com.jon.common.utils.Key;
import com.jon.common.utils.PrefUtils;
import com.jon.common.CotApplication;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.ConcurrentModificationException;
import java.util.List;
import java.util.Locale;
import java.util.PrimitiveIterator;
import java.util.Random;
import java.util.concurrent.TimeUnit;
import java.util.stream.DoubleStream;

public class GeneratorCotFactory extends CotFactory {
    private Random random = new Random();

    private List<IconData> icons;
    private List<String> callsigns;
    private int iconCount;
    private double movementSpeed;
    private double travelDistance;
    private double distributionRadius;
    private boolean followGps;
    private double centreLat;
    private double centreLon;
    private boolean stayAtGroundLevel;
    private double centreAlt;
    private long staleTimer;
    private Point distributionCentre;

    private static class IconData {
        CursorOnTarget cot;
        Point.Offset offset;
        IconData(CursorOnTarget cot, Point.Offset offset) {
            this.cot = cot;
            this.offset = offset;
        }
    }

    GeneratorCotFactory(SharedPreferences prefs) {
        super(prefs);
        iconCount = PrefUtils.parseInt(prefs, Key.ICON_COUNT);
        callsigns = getCallsigns();
        distributionRadius = PrefUtils.parseDouble(prefs, Key.RADIAL_DISTRIBUTION);
        movementSpeed = PrefUtils.parseDouble(prefs, Key.MOVEMENT_SPEED) * Constants.MPH_TO_METRES_PER_SECOND;
        followGps = PrefUtils.getBoolean(prefs, Key.FOLLOW_GPS_LOCATION);
        centreLat = PrefUtils.parseDouble(prefs, Key.CENTRE_LATITUDE);
        centreLon = PrefUtils.parseDouble(prefs, Key.CENTRE_LONGITUDE);
        stayAtGroundLevel = PrefUtils.getBoolean(prefs, Key.STAY_AT_GROUND_LEVEL);
        centreAlt = stayAtGroundLevel ? 0.0 : (double)PrefUtils.getInt(prefs, Key.CENTRE_ALTITUDE);
        staleTimer = PrefUtils.getInt(prefs, Key.STALE_TIMER);

        /* Stop any fuckery with distribution radii */
        movementSpeed = Math.min(movementSpeed, distributionRadius/2.0);
        travelDistance = movementSpeed * PrefUtils.getInt(prefs, Key.TRANSMISSION_PERIOD);
    }

    @Override
    protected void clear() {
        if (icons != null) {
            icons.clear();
            icons = null;
        }
    }

    @Override
    protected List<CursorOnTarget> generate() {
        try {
            return (icons == null) ? initialise() : update();
        } catch (ConcurrentModificationException e) {
            return new ArrayList<>();
        }
    }

    @Override
    protected List<CursorOnTarget> initialise() {
        icons = new ArrayList<>();
        updateDistributionCentre();
        final UtcTimestamp now = UtcTimestamp.now();
        PrimitiveIterator.OfDouble distanceItr = weightedRadialIterator();
        PrimitiveIterator.OfDouble courseItr = randomIterator(0.0, 360.0);
        PrimitiveIterator.OfDouble altitudeItr = randomIterator(centreAlt-distributionRadius, centreAlt+distributionRadius);
        for (int i = 0; i < iconCount; i++) {
            CursorOnTarget cot = new CursorOnTarget();
            cot.uid = String.format(Locale.ENGLISH, "%s_%04d", DeviceUid.get(), i);
            cot.callsign = callsigns.get(i);
            cot.time = cot.start = now;
            cot.setStaleDiff(staleTimer, TimeUnit.MINUTES);
            cot.team = CotTeam.fromPrefs(prefs);
            cot.role = CotRole.fromPrefs(prefs);
            cot.speed = movementSpeed;
            cot.lat = distributionCentre.lat * Constants.RAD_TO_DEG;
            cot.lon = distributionCentre.lon * Constants.RAD_TO_DEG;
            cot.hae = initialiseAltitude(altitudeItr);
            cot.battery = battery.getPercentage();
            Point.Offset initialOffset = generateInitialOffset(distanceItr, courseItr);
            setPositionFromOffset(cot, initialOffset);
            cot.course = initialOffset.theta;
            icons.add(new IconData(cot, initialOffset));
        }
        return getCot();
    }

    protected List<CursorOnTarget> update() {
        updateDistributionCentre();
        final UtcTimestamp now = UtcTimestamp.now();
        PrimitiveIterator.OfDouble courseItr = randomIterator(0.0, 360.0);
        for (IconData icon : icons) {
            icon.cot.time = icon.cot.start = now;
            icon.cot.setStaleDiff(staleTimer, TimeUnit.MINUTES);
            icon.offset = generateBoundedOffset(courseItr, Point.fromCot(icon.cot));
            Point oldPoint = Point.fromCot(icon.cot);
            setPositionFromOffset(icon.cot, icon.offset);
            icon.cot.course = bearing(oldPoint, Point.fromCot(icon.cot));
            icon.cot.hae = updateAltitude(icon.cot.hae);
            icon.cot.battery = battery.getPercentage();
        }
        return getCot();
    }

    private List<String> getCallsigns() {
        List<String> callsigns = new ArrayList<>();
        if (PrefUtils.getBoolean(prefs, Key.RANDOM_CALLSIGNS)) {
            /* Grab the list of all valid callsigns and shuffle it into a random order */
            Context context = CotApplication.getContext();
            List<String> allCallsigns = Arrays.asList(context.getResources().getStringArray(R.array.atakCallsigns));
            Collections.shuffle(allCallsigns);
            /* Extract some at random */
            for (int i = 0; i < iconCount; i++) {
                callsigns.add(allCallsigns.get(i % allCallsigns.size())); // modulus, just in case iconCount > allCallsigns.size
            }
        } else {
            /* Use custom callsign as entered in the settings */
            final String baseCallsign = PrefUtils.getString(prefs, Key.CALLSIGN);
            for (int i = 0; i < iconCount; i++) {
                callsigns.add(String.format(Locale.ENGLISH, "%s-%d", baseCallsign, i));
            }
        }
        return callsigns;
    }

    private void setPositionFromOffset(CursorOnTarget cot, Point.Offset newOffset) {
        Point location = Point.fromCot(cot).add(newOffset);
        cot.lat = location.lat * Constants.RAD_TO_DEG;
        cot.lon = location.lon * Constants.RAD_TO_DEG;
    }

    private void updateDistributionCentre() {
        distributionCentre = new Point(
                getCentreLatitudeDegrees() * Constants.DEG_TO_RAD,
                getCentreLongitudeDegrees() * Constants.DEG_TO_RAD
        );
    }

    private Point.Offset generateInitialOffset(PrimitiveIterator.OfDouble distanceItr, PrimitiveIterator.OfDouble courseItr) {
        return new Point.Offset(distanceItr.next(), courseItr.next());
    }

    private Point.Offset generateBoundedOffset(final PrimitiveIterator.OfDouble courseItr, final Point startPoint) {
        final Point.Offset offset = new Point.Offset(travelDistance, courseItr.next());
        final Point endPoint = startPoint.add(offset);
        if (arcdistance(endPoint, distributionCentre) > distributionRadius) {
            /* Invalid offset, so try again */
            return generateBoundedOffset(courseItr, startPoint);
        } else {
            return offset;
        }
    }

    private double arcdistance(final Point p1, final Point p2) {
        final double lat1 = p1.lat;
        final double lat2 = p2.lat;
        final double dlat = p2.lat - p1.lat;
        final double dlon = p2.lon - p1.lon;
        /* I can feel myself getting sweaty just looking at this */
        final double a = Math.sin(dlat / 2.0) * Math.sin(dlat / 2.0) + Math.cos(lat1) * Math.cos(lat2) * Math.sin(dlon / 2.0) * Math.sin(dlon / 2.0);
        return 2.0 * Constants.EARTH_RADIUS * Math.atan2(Math.sqrt(a), Math.sqrt(1.0 - a));
    }

    private double bearing(Point start, Point end) {
        final double y = Math.sin(end.lon-start.lon) * Math.cos(end.lat);
        final double x = Math.cos(start.lat) * Math.sin(end.lat) - Math.sin(start.lat) * Math.cos(end.lat) * Math.cos(end.lon-start.lon);
        return ( Math.atan2(y, x) * Constants.RAD_TO_DEG + 360.0 ) % 360.0;
    }

    private PrimitiveIterator.OfDouble randomIterator(double dx) {
        return randomIterator(-dx, dx);
    }

    private PrimitiveIterator.OfDouble randomIterator(double min, double max) {
        return random.doubles(min, max).iterator();
    }

    private PrimitiveIterator.OfDouble weightedRadialIterator() {
        final double bound = Math.pow(distributionRadius, 2.0);
        /* p(x) proportional to sqrt(x), hopefully?  */
        return DoubleStream.generate(() -> Math.sqrt(random.nextDouble() * bound)).iterator();
    }

    private double getCentreLatitudeDegrees() {
        if (followGps) {
            return gpsCoords.latitude();
        } else {
            return centreLat;
        }
    }

    private double getCentreLongitudeDegrees() {
        if (followGps) {
            return gpsCoords.longitude();
        } else {
            return centreLon;
        }
    }

    private List<CursorOnTarget> getCot() {
        List<CursorOnTarget> crumbs = new ArrayList<>();
        for (IconData iconData : icons) {
            crumbs.add(iconData.cot);
        }
        return crumbs;
    }

    private double initialiseAltitude(PrimitiveIterator.OfDouble altitudeIterator) {
        if (stayAtGroundLevel) {
            return 0.0;
        } else {
            /* Can't have altitude below 0 */
            return Math.max(0.0, altitudeIterator.next());
        }
    }

    private double updateAltitude(double altitude) {
        if (stayAtGroundLevel) {
            return 0.0;
        } else {
            /* Direction is either -1, 0 or +1; representing falling, staying steady or rising */
            final int direction = random.ints(1, -1, 1).iterator().next();
            double newAltitude = altitude + (direction * movementSpeed);

            /* Not going below ground */
            if (newAltitude < 0.0) newAltitude = 0.0;

            /* Clip within the bounds of the distribution radius*/
            if (newAltitude < centreAlt-distributionRadius) newAltitude = centreAlt-distributionRadius;
            if (newAltitude > centreAlt+distributionRadius) newAltitude = centreAlt+distributionRadius;

            return newAltitude;
        }
    }
}

File: cotgenerator/src/main/java/com/jon/cotgenerator/GeneratorCotFactory.kt
package com.jon.cotgenerator

import android.content.SharedPreferences
import com.jon.common.CotApplication
import com.jon.common.cot.CotRole
import com.jon.common.cot.CotTeam
import com.jon.common.cot.CursorOnTarget
import com.jon.common.cot.UtcTimestamp
import com.jon.common.prefs.*
import com.jon.common.service.CotFactory
import com.jon.common.service.Offset
import com.jon.common.service.Point
import com.jon.common.utils.Bearing
import com.jon.common.utils.Constants
import com.jon.common.utils.GeometryUtils.arcdistance
import com.jon.cotgenerator.streams.DoubleRandomStream
import com.jon.cotgenerator.streams.IntRandomStream
import com.jon.cotgenerator.streams.RadialDistanceRandomStream
import com.jon.cotgenerator.streams.RandomStream
import java.util.*
import java.util.concurrent.TimeUnit
import kotlin.math.max
import kotlin.math.min

internal class GeneratorCotFactory(prefs: SharedPreferences) : CotFactory(prefs) {
    private data class IconData(var cot: CursorOnTarget, var offset: Offset)

    private val random = Random(System.currentTimeMillis())

    private val useRandomCallsigns = prefs.getBooleanFromPair(GeneratorPrefs.RANDOM_CALLSIGNS)
    private val useRandomTeams = prefs.getBooleanFromPair(GeneratorPrefs.RANDOM_COLOUR)
    private val useRandomRoles = prefs.getBooleanFromPair(GeneratorPrefs.RANDOM_ROLE)
    private val iconCount = prefs.parseIntFromPair(GeneratorPrefs.ICON_COUNT)
    private val callsigns = getCallsigns()
    private val distributionRadius = prefs.parseDoubleFromPair(GeneratorPrefs.RADIAL_DISTRIBUTION)
    private val followGps = prefs.getBooleanFromPair(GeneratorPrefs.FOLLOW_GPS_LOCATION)
    private val centreLat = prefs.parseDoubleFromPair(GeneratorPrefs.CENTRE_LATITUDE)
    private val centreLon = prefs.parseDoubleFromPair(GeneratorPrefs.CENTRE_LONGITUDE)
    private val stayAtGroundLevel = prefs.getBooleanFromPair(GeneratorPrefs.STAY_AT_GROUND_LEVEL)
    private val centreAlt = if (stayAtGroundLevel) 0.0 else prefs.parseDoubleFromPair(GeneratorPrefs.CENTRE_ALTITUDE)
    private val staleTimer = prefs.getIntFromPair(CommonPrefs.STALE_TIMER).toLong()
    private var movementSpeed = prefs.parseDoubleFromPair(GeneratorPrefs.MOVEMENT_SPEED) * Constants.MPH_TO_METRES_PER_SECOND
    private var travelDistance: Double

    private lateinit var distributionCentre: Point
    private var icons = mutableListOf<IconData>()

    init {
        /* Stop any fuckery with distribution radii */
        movementSpeed = min(movementSpeed, distributionRadius / 2.0)
        travelDistance = movementSpeed * prefs.getIntFromPair(CommonPrefs.TRANSMISSION_PERIOD)
    }

    override fun clear() {
        icons.clear()
    }

    override fun generate(): List<CursorOnTarget> {
        return try {
            if (icons.isEmpty()) initialise() else update()
        } catch (e: ConcurrentModificationException) {
            ArrayList()
        }
    }

    override fun initialise(): List<CursorOnTarget> {
        icons = ArrayList()
        updateDistributionCentre()
        val now = UtcTimestamp.now()
        val distanceItr = weightedRadialIterator()
        val courseItr = doubleIterator(0.0, 360.0)
        val altitudeItr = doubleIterator(centreAlt - distributionRadius, centreAlt + distributionRadius)
        for (i in 0 until iconCount) {
            val cot = CursorOnTarget()
            cot.uid = "%s_%04d".format(deviceUidRepository.getUid(), i)
            cot.callsign = callsigns[i]
            cot.start = now
            cot.time = cot.start
            cot.setStaleDiff(staleTimer, TimeUnit.MINUTES)
            cot.team = CotTeam.fromPrefs(prefs, useRandomTeams)
            cot.role = CotRole.fromPrefs(prefs, useRandomRoles)
            cot.speed = movementSpeed
            cot.lat = distributionCentre.lat * Constants.RAD_TO_DEG
            cot.lon = distributionCentre.lon * Constants.RAD_TO_DEG
            cot.hae = initialiseAltitude(altitudeItr)
            cot.battery = batteryRepository.getPercentage()
            val initialOffset = generateInitialOffset(distanceItr, courseItr)
            setPositionFromOffset(cot, initialOffset)
            cot.course = initialOffset.theta
            icons.add(IconData(cot, initialOffset))
        }
        return getCotList()
    }

    override fun update(): List<CursorOnTarget> {
        updateDistributionCentre()
        val now = UtcTimestamp.now()
        val courseItr = doubleIterator(0.0, 360.0)
        icons.forEach {
            it.cot.start = now
            it.cot.time = now
            it.cot.setStaleDiff(staleTimer, TimeUnit.MINUTES)
            it.offset = generateBoundedOffset(courseItr, Point.fromCot(it.cot))
            val oldPoint = Point.fromCot(it.cot)
            setPositionFromOffset(it.cot, it.offset)
            val newPoint = Point.fromCot(it.cot)
            it.cot.course = Bearing.from(oldPoint).to(newPoint)
            it.cot.hae = updateAltitude(it.cot.hae)
            it.cot.battery = batteryRepository.getPercentage()
        }
        return getCotList()
    }

    private fun getCallsigns(): List<String> {
        val callsigns: MutableList<String> = ArrayList()
        if (useRandomCallsigns) {
            /* Grab the list of all valid callsigns and shuffle it into a random order */
            val resources = CotApplication.context.resources
            val allCallsigns = mutableListOf(*resources.getStringArray(R.array.atakCallsigns))
            allCallsigns.shuffle()
            /* Extract some at random */
            for (i in 0 until iconCount) {
                callsigns.add(allCallsigns[i % allCallsigns.size]) // modulus, just in case iconCount > allCallsigns.size
            }
        } else {
            /* Use custom callsign as entered in the settings */
            val baseCallsign = prefs.getStringFromPair(CommonPrefs.CALLSIGN)
            for (i in 0 until iconCount) {
                callsigns.add(String.format(Locale.ENGLISH, "%s-%d", baseCallsign, i))
            }
        }
        return callsigns
    }

    private fun setPositionFromOffset(cot: CursorOnTarget, newOffset: Offset) {
        val (lat, lon) = Point.fromCot(cot).add(newOffset)
        cot.lat = lat * Constants.RAD_TO_DEG
        cot.lon = lon * Constants.RAD_TO_DEG
    }

    private fun updateDistributionCentre() {
        distributionCentre = Point(
                lat = centreLatitudeDegrees() * Constants.DEG_TO_RAD,
                lon = centreLongitudeDegrees() * Constants.DEG_TO_RAD
        )
    }

    private fun generateInitialOffset(distanceItr: RandomStream<Double>, courseItr: RandomStream<Double>): Offset {
        return Offset(
                R = distanceItr.next(),
                theta = courseItr.next()
        )
    }

    private fun generateBoundedOffset(
            courseItr: RandomStream<Double>,
            startPoint: Point,
            attemptsRemaining: Int = MAX_OFFSET_GENERATION_ATTEMPTS
    ): Offset {
        if (attemptsRemaining == 0) {
            /* The recursive algorithm has failed too many times, so to avoid a stack overflow we just pick
             * a random point along the radius of the centrepoint and generate an Offset that will take us there */
            val offsetToEdge = Offset(travelDistance, courseItr.next())
            val nextPoint = distributionCentre.add(offsetToEdge)
            return Offset.from(startPoint).to(nextPoint)
        } else {
            val offset = Offset(travelDistance, courseItr.next())
            val endPoint = startPoint.add(offset)
            return if (arcdistance(endPoint, distributionCentre) > distributionRadius) {
                /* Invalid offset, so try again */
                generateBoundedOffset(courseItr, startPoint, attemptsRemaining - 1)
            } else {
                offset
            }
        }
    }

    private fun doubleIterator(min: Double, max: Double): RandomStream<Double> {
        return DoubleRandomStream(random, min, max)
    }

    private fun weightedRadialIterator(): RandomStream<Double> {
        return RadialDistanceRandomStream(random, distributionRadius)
    }

    private fun centreLatitudeDegrees() = if (followGps) {
        gpsRepository.latitude()
    } else {
        centreLat
    }

    private fun centreLongitudeDegrees() = if (followGps) {
        gpsRepository.longitude()
    } else {
        centreLon
    }

    private fun getCotList(): List<CursorOnTarget> {
        return icons.map { it.cot }
    }

    private fun initialiseAltitude(altitudeIterator: RandomStream<Double>): Double {
        return if (stayAtGroundLevel) {
            0.0
        } else {
            /* Can't have altitude below 0 */
            max(0.0, altitudeIterator.next())
        }
    }

    private fun updateAltitude(altitude: Double): Double {
        return if (stayAtGroundLevel) {
            0.0
        } else {
            /* Direction is either -1, 0 or +1; representing falling, staying steady or rising respectively */
            val direction = IntRandomStream(random, -1, 1).next()
            var newAltitude = altitude + direction * movementSpeed

            /* Not going below ground */
            if (newAltitude < 0.0) newAltitude = 0.0

            /* Clip within the bounds of the distribution radius*/
            if (newAltitude < centreAlt - distributionRadius) newAltitude = centreAlt - distributionRadius
            if (newAltitude > centreAlt + distributionRadius) newAltitude = centreAlt + distributionRadius
            newAltitude
        }
    }

    private companion object {
        const val MAX_OFFSET_GENERATION_ATTEMPTS = 10
    }
}
